### DISCUSSION TITLE
How do we deal with the pubspec.lock file?

### DISCUSSION THREAD

**Discussion by JesseRiemens** [meta-flutter/meta-flutter#477] [Q&A]
The pubspec.lock file (as explained in flutter documentation) is the file that should enable reproductive builds, since getting packages with `flutter pub get --enforce-lockfile` will get packages using the resolved dependencies in the lockfile. 

It looks like we remove the lockfile in `conf/include/common.inc`. How do I then know that builds from the same commit hash are reproducible?
(This was added quite recently: [00994a](https://github.com/meta-flutter/meta-flutter/commit/00994a2d957626895271193167bc0f782bbb8df4))

**Comment by jwinarske** [comment]
@JesseRiemens reproducibility is relative to the point in time when the pubcache archive is created, which is 100% relative to Flutter SDK version. 

I understand if you want to pin to set of versions, but that breaks rolling the Flutter SDK pretty quick.  Which becomes a maintenance nightmare; and likely removal of many apps from the layer (largely a third party problem).

My plan for a pubspec fetcher (similar pattern to Yocto rust) forces resolution at recipe generation time.  Which in my view is the best option.  

The same problem exists with rust lock files.  In which case one of the steps of creating a recipe specific to a toolchain (same idea as flutter SDK) involves first removing the caro.lock file.  Which then you turn the recipe into the virtual lock file.  In our case (today) is the pubspec archive file.  So in the current state of things your build is reproducible relative to the artifacts in your DL_DIR.  

The better solution will be a custom pubspec fetcher; yet it's not clear yet how brittle that would be.

Being pedantically correct without a custom pubspec fetcher today equates to excessive maintenance overhead.  The goal by removing the lock file was to make rolling an SDK easy to do with 100+ recipes.

**Reply by JesseRiemens** [reply]
Thanks for your reply!
If I understand you correctly, there is reproducibility if the same flutter SDK is used for two concurrent builds, when using the same DL_DIR. 

Does that mean that if I build from the exact same recipe set on two different machines, not sharing DL_DIR, it is not guaranteed to contain the same dependencies? 

And would that be solved by using a custom pubspec fetcher?

**Comment by jwinarske** [comment]
> Does that mean that if I build from the exact same recipe set on two different machines, not sharing DL_DIR, it is not guaranteed to contain the same dependencies?
> 

That only depends on the SDK version.  If the SDK is constant it will be the same regardless of DL_DIR

> And would that be solved by using a custom pubspec fetcher?

Yes.  pubspec fetcher is currently under development.  I figure a week or two until it's pushed.
